_posts/2019-01-18-monorepos-aws-codebuild.markdown

```diff
@@ -0,0 +1,64 @@
+
+The first step was to introduce something to help us manage the monorepo. I had previously used 'lerna' to manage another node-based monorepo with some success. This began a reasonably lengthy task to restructure what he already had. This meant moving from a heavily nested module structure, to a more flattened structure - which is technically what would have happened had each module been seperated into its own repository. With this in-place, it was now a little easier to manage the dependencies of each module and enable a clear seperation of concerns. We were also using a feature in lerna called 'hoisting' which deduplicates dependencies that many packages may rely on.
+
+Unfortunately, we had a small issue with hoisting which meant that we had to remove it. Hoisting involves installing packages in the base node_modules directory of the repository as opposed to the specific package - your 'child' packages thereby resolve all their dependencies at the base as opposed to their own node_modules folders. However, A few of our packages needed to bundle their dependencies, and this was unfortunately impossible to do with lerna's hoisting, because they would attempt to package their local node_modules folders which contained nothing. Had lerna's hoisting had the ability to be limited to development dependencies, this issue may have gone away. But alas, it did not, and therefore we needed to disable hoisting.
+
+Despite the movement to lerna to better manage the repository, this meant that we still had to reckon with fairly large build times which were agitated by the removal of hoisting. It was at this point I started inspecting our build environment and actions to determine possible improvements. This involved learning a lot more as to how npm packaging, caching and installation works, and from this I was able to come up with a few things to try out.
+
+The next improvement to be made was to use `npm ci`. This instructs npm to not resolve dependencies from directly from definitions in the package.json file, and instead use the package-lock.json file. Fundamentally, this instructs npm to install all packages called out in the lock file, rather than resolving everything from the top level dependencies. This provided a reasonably modest speed improvement (roughly ~20%), which is not bad for what was a minor one-line change.
```

Comment:
This instructs npm to not resolve dependencies from directly from definitions in the package.json file -> 
extra from